### PR TITLE
Small fixes for group_by_use_nulls

### LINKED ISSUE #60538: Logical error 'Unexpected return type from equals' in group_by_use_nulls + GROUPING SETS
https://fiddle.clickhouse.com/dba51b74-4843-46e5-abbc-8cb5d0310373

And it works on `latest`, though. But broken on `head`

```sql

DROP TABLE IF EXISTS test_grouping_sets_predicate;

CREATE TABLE test_grouping_sets_predicate
(
    day_ Date,
    type_1 String
)
ENGINE=MergeTree
ORDER BY day_;

INSERT INTO test_grouping_sets_predicate SELECT toDate('2023-01-05') AS day_, 'hello, world' FROM numbers (10);

SET group_by_use_nulls = true;

SELECT *
FROM
( SELECT
    day_,
    type_1
FROM test_grouping_sets_predicate
GROUP BY
    GROUPING SETS ( (day_, type_1), (day_) )
)
WHERE day_ = '2023-01-05';

```

```
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Unexpected return type from equals. Expected Nullable. Got UInt8: while executing 'FUNCTION equals(day_ : 0, '2023-01-05' :: 1) -> equals(day_, '2023-01-05') Nullable(UInt8) : 3'. (LOGICAL_ERROR)
```

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix logical errors in queries with `GROUPING SETS` and `WHERE` and `group_by_use_nulls = true`, close https://github.com/ClickHouse/ClickHouse/issues/60538

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 26846bd238813c56ad36654ce6dac4b57e17e0e1 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63405/26846bd238813c56ad36654ce6dac4b57e17e0e1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 1]
```diff
@@ -4815,6 +4815,19 @@ ProjectionNames QueryAnalyzer::resolveMatcher(QueryTreeNodePtr & matcher_node, I
...
+        for (auto & [node, _] : matched_expression_nodes_with_names)
+        {
+            auto it = scope.nullable_group_by_keys.find(node);
+            if (it != scope.nullable_group_by_keys.end())
+            {
+                node = it->node->clone();
+                node->convertToNullable();
+            }
+        }
+    }
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:4829]: Hmm. Interesting case.
Is it about something like `SELECT ... GROUP BY ... ORDER BY *`?

File: src/Processors/QueryPlan/Optimizations/filterPushDown.cpp [hunk 2]
```diff
@@ -428,6 +428,9 @@ size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes
         /// of the grouping sets, we could not push the filter down.
         if (aggregating->isGroupingSets())
         {
+            /// Cannot push down filter if type has been changed.
+            if (aggregating->isGroupByUseNulls())
+                return 0;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/filterPushDown.cpp:433]: Makes sense. But it sad, that `group_by_use_nulls` limits optimizations so much..Do you think it's possible to update the types in the DAG?

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 1 above)
Comment (vdimir) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:4829]: Should be the same, we resolve `*` using the same method not depending on it's source as usual identifiers. Added tests for it

File: src/Processors/QueryPlan/Optimizations/filterPushDown.cpp (hunk 2 above)
Comment (vdimir) [reply src/Processors/QueryPlan/Optimizations/filterPushDown.cpp:433]: Probably yes, but I cannot come up with solution from the top of my head

File: src/Analyzer/ValidationUtils.cpp [hunk 3]
```diff
@@ -26,6 +26,10 @@ namespace
 
 void validateFilter(const QueryTreeNodePtr & filter_node, std::string_view exception_place_message, const QueryTreeNodePtr & query_node)
 {
+    if (filter_node->getNodeType() == QueryTreeNodeType::LIST)
```
Comment (vdimir) [inline src/Analyzer/ValidationUtils.cpp:29]: Now error is better, I will adjust the test https://s3.amazonaws.com/clickhouse-test-reports/63405/c67a43e3a7354c3db488df44c05e5a0d56a454c5/fast_test.html

```
2024-05-07 16:51:58 Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Unsupported expression '*' in filter. (BAD_ARGUMENTS)
2024-05-07 16:51:58 (query: SELECT * WHERE *;  -- { serverError UNSUPPORTED_METHOD })
```

**Pull request merged**